neptune/hierarchy_test.go

```diff
@@ -2,13 +2,363 @@ package neptune
+	testCodes         = []string{"cpih1dim1S90401", "cpih1dim1S90402"}
+	testIds           = []string{"cpih1dim1aggid--cpih1dim1S90401", "cpih1dim1aggid--cpih1dim1S90402"}
+	testAllIds        = []string{"cpih1dim1aggid--cpih1dim1S90401", "cpih1dim1aggid--cpih1dim1S90402",
+		"cpih1dim1aggid--cpih1dim1G90400", "cpih1dim1aggid--cpih1dim1G90400",
+		"cpih1dim1aggid--cpih1dim1T90000", "cpih1dim1aggid--cpih1dim1T90000",
+		"cpih1dim1aggid--cpih1dim1A0", "cpih1dim1aggid--cpih1dim1A0"}
+	testClonedIds = []string{
+		"62bab579-e923-7cb2-3be0-34d09dc0567b",
+		"acbab579-e923-87df-e59a-9daf2ffed388",
+		"b6bab57a-604d-8a7f-59f5-1d496c9b3ca5",
+		"08bab57a-604d-9cd9-492f-e879cee05502",
+		"6cbab57a-604d-f176-9370-c60c19369801",
+	}
+)
+
+func TestNeptuneDB_GetCodesWithData(t *testing.T) {
+
+	Convey("Given a mocked neptune DB that returns a code list", t, func() {
+		poolMock := &internal.NeptunePoolMock{
+			GetStringListFunc: internal.ReturnCodesList,
+		}
+		db := mockDB(poolMock)
+
+		Convey("When GetCodesWithData is called", func() {
+			codes, err := db.GetCodesWithData(ctx, testAttempt, testInstanceID, testDimensionName)
+
+			Convey("Then no error is returned", func() {
+				So(err, ShouldBeNil)
+			})
+
+			Convey("Then the expected list of codes is returned", func() {
+				So(len(codes), ShouldEqual, 2)
+				So(codes, ShouldContain, "cpih1dim1S90401")
+				So(codes, ShouldContain, "cpih1dim1S90402")
+				expectedQuery := `g.V().hasLabel('_f0a2f3f2-cc86-4bbb-a549-ffc99c89292c_aggregate').values('value')`
+				So(len(poolMock.GetStringListCalls()), ShouldEqual, 1)
+				So(poolMock.GetStringListCalls()[0].Query, ShouldEqual, expectedQuery)
+			})
+		})
+	})
+}
+
+func TestNeptuneDB_GetGenericHierarchyNodeIDs(t *testing.T) {
+
+	Convey("Given a mocked neptune DB that returns a list of generic hierarchy node IDs (leaves)", t, func() {
+		poolMock := &internal.NeptunePoolMock{
+			GetStringListFunc: internal.ReturnGenericHierarchyLeavesIDs,
+		}
+		db := mockDB(poolMock)
+
+		Convey("When GetGenericHierarchyNodeIDs is called", func() {
+			ids, err := db.GetGenericHierarchyNodeIDs(ctx, testAttempt, testCodeListID, testCodes)
+
+			Convey("Then no error is returned", func() {
+				So(err, ShouldBeNil)
+			})
+
+			Convey("Then the expected list of IDs is returned and the expected query is executed, in any order of IDs", func() {
+				So(len(ids), ShouldEqual, 2)
+				So(ids, ShouldContain, "cpih1dim1aggid--cpih1dim1S90401")
+				So(ids, ShouldContain, "cpih1dim1aggid--cpih1dim1S90402")
+				expectedQueryOp1 := `g.V().hasLabel('_generic_hierarchy_node_cpih1dim1aggid').has('code',within(['cpih1dim1S90401','cpih1dim1S90402'])).id()`
+				expectedQueryOp2 := `g.V().hasLabel('_generic_hierarchy_node_cpih1dim1aggid').has('code',within(['cpih1dim1S90402','cpih1dim1S90401'])).id()`
+				So(len(poolMock.GetStringListCalls()), ShouldEqual, 1)
+				So(poolMock.GetStringListCalls()[0].Query, ShouldBeIn, []string{expectedQueryOp1, expectedQueryOp2})
```

Comment:
I think you are right - I tried to run the test multiple times, expecting only query1, and it always worked.

Reply:
TestNeptuneDB_CloneNodes failed randomly if I only expected one of the queries. As we don't care about the order (any traversal is independent of the other) I think I will leave it with the different options.

Reply:
I think it is random because the results are a map. I came across the same issue in my tests - because the results are a map, go does no guarantee the iteration order of map elements.